java/src/main/native/src/row_conversion.cu

```diff
@@ -594,134 +594,106 @@ __global__ void copy_to_rows(const size_type num_rows, const size_type num_colum
   // This has been broken up for us in the tile_info struct, so we don't have
   // any calculation to do here, but it is important to note.
 
-  constexpr unsigned stages_count = NUM_TILES_PER_KERNEL_LOADED;
   auto group = cooperative_groups::this_thread_block();
+  auto warp = cooperative_groups::tiled_partition<NUM_THREADS_IN_WARP>(group);
   extern __shared__ int8_t shared_data[];
-  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_tile};
 
 #ifdef ASYNC_MEMCPY_SUPPORTED
-  __shared__ cuda::barrier<cuda::thread_scope_block> tile_barrier[NUM_TILES_PER_KERNEL_LOADED];
+  __shared__ cuda::barrier<cuda::thread_scope_block> tile_barrier;
   if (group.thread_rank() == 0) {
-    for (int i = 0; i < NUM_TILES_PER_KERNEL_LOADED; ++i) {
-      init(&tile_barrier[i], group.size());
-    }
+    init(&tile_barrier, group.size());
   }
   group.sync();
 #endif // ASYNC_MEMCPY_SUPPORTED
 
-  auto const tiles_remaining =
-      std::min(static_cast<uint>(tile_infos.size()) - blockIdx.x * NUM_TILES_PER_KERNEL_TO_ROWS,
-               static_cast<uint>(NUM_TILES_PER_KERNEL_TO_ROWS));
+  auto const tile = tile_infos[blockIdx.x];
+  auto const num_tile_cols = tile.num_cols();
+  auto const num_tile_rows = tile.num_rows();
+  auto const tile_row_size = tile.get_shared_row_size(col_offsets, col_sizes);
+  auto const starting_column_offset = col_offsets[tile.start_col];
+
+  // to do the copy we need to do n column copies followed by m element copies OR
+  // we have to do m element copies followed by r row copies. When going from column
+  // to row it is much easier to copy by elements first otherwise we would need a running
+  // total of the column sizes for our tile, which isn't readily available. This makes it
+  // more appealing to copy element-wise from input data into shared matching the end layout
+  // and do row-based memcopies out.
+
+  // read each column across the tile
+  // each warp takes a column with each thread of a warp taking a row
+  for (int relative_col = warp.meta_group_rank(); relative_col < num_tile_cols;
+       relative_col += warp.meta_group_size()) {
+    for (int relative_row = warp.thread_rank(); relative_row < num_tile_rows;
```

Comment:
Question for those more knowledgeable that myself: In spite of `warp.size()` being a `constexpr`, we would not use `#pragma unroll` here because the body of the inner `for` loop is long, right?

Reply:
With the conditionals inside the loop and the size of the loop I wouldn't expect it to improve things to attempt and unroll it, but I honestly don't know if it would. It comes may depend on how large `num_tile_rows/warp.size()` is as that is how many iterations each thread will do.